src/tactic/norm_num.lean

```diff
@@ -36,270 +55,1100 @@ variable {α : Type u}
 lemma subst_into_neg {α} [has_neg α] (a ta t : α) (pra : a = ta) (prt : -ta = t) : -a = t :=
 by simp [pra, prt]
 
-theorem bit0_zero [add_group α] : bit0 (0 : α) = 0 := add_zero _
+/-- The result type of `match_numeral`, either `0`, `1`, or a top level
+decomposition of `bit0 e` or `bit1 e`. The `other` case means it is not a numeral. -/
+meta inductive match_numeral_result
+| zero | one | bit0 (e : expr) | bit1 (e : expr) | other
+
+/-- Unfold the top level constructor of the numeral expression. -/
+meta def match_numeral : expr → match_numeral_result
+| `(bit0 %%e) := match_numeral_result.bit0 e
+| `(bit1 %%e) := match_numeral_result.bit1 e
+| `(@has_zero.zero _ _) := match_numeral_result.zero
+| `(@has_one.one _ _) := match_numeral_result.one
+| _ := match_numeral_result.other
 
-theorem bit1_zero [add_group α] [has_one α] : bit1 (0 : α) = 1 :=
-by rw [bit1, bit0_zero, zero_add]
+theorem zero_succ {α} [semiring α] : (0 + 1 : α) = 1 := zero_add _
+theorem one_succ {α} [semiring α] : (1 + 1 : α) = 2 := rfl
+theorem bit0_succ {α} [semiring α] (a : α) : bit0 a + 1 = bit1 a := rfl
+theorem bit1_succ {α} [semiring α] (a b : α) (h : a + 1 = b) : bit1 a + 1 = bit0 b :=
+h ▸ by simp [bit1, bit0, add_left_comm]
 
-lemma pow_bit0_helper [monoid α] (a t : α) (b : ℕ) (h : a ^ b = t) :
-  a ^ bit0 b = t * t :=
-by simp [pow_bit0, h]
+section
+open match_numeral_result
 
-lemma pow_bit1_helper [monoid α] (a t : α) (b : ℕ) (h : a ^ b = t) :
-  a ^ bit1 b = t * t * a :=
-by simp [pow_bit1, h]
+/-- Given `a`, `b` natural numerals, proves `⊢ a + 1 = b`, assuming that this is provable.
+(It may prove garbage instead of failing if `a + 1 = b` is false.) -/
+meta def prove_succ : instance_cache → expr → expr → tactic (instance_cache × expr)
+| c e r := match match_numeral e with
+  | zero := c.mk_app ``zero_succ []
+  | one := c.mk_app ``one_succ []
+  | bit0 e := c.mk_app ``bit0_succ [e]
+  | bit1 e := do
+    let r := r.app_arg,
+    (c, p) ← prove_succ c e r,
+    c.mk_app ``bit1_succ [e, r, p]
+  | _ := failed
+  end
+end
 
-lemma lt_add_of_pos_helper [ordered_cancel_add_comm_monoid α]
-  (a b c : α) (h : a + b = c) (h₂ : 0 < b) : a < c :=
-h ▸ (lt_add_iff_pos_right _).2 h₂
+theorem zero_adc {α} [semiring α] (a b : α) (h : a + 1 = b) : 0 + a + 1 = b := by rwa zero_add
```

Comment:
The stuff about addition is true in any `add_comm_monoid`, right?

Reply:
The constraint on these theorems is that they have to use only one typeclass for `instance_cache.mk_app` to work properly, and we would need both `add_comm_monoid` and `has_one` to make sense of these. I don't know any examples of `add_comm_monoid`s with one which are not semirings, but we have three options:

1. Use `add_comm_monoid` and `has_one` anyway, modify the tactic as appropriate
2. Add `add_comm_monoid_with_one` typeclass and use that
3. Use the weakest typeclass that implies both `add_comm_monoid` and `has_one`, which is currently `semiring`

While (1) is our default position, it tends to generate quite large terms (if you have ever looked at a theorem using `coe : rat -> real` with `pp.all` you will know what I mean) and I regard this as a bit of a mistake, as we almost never make use of this generality and it has significant performance costs. (2) is an option but I don't like the idea of extending the typeclass hierarchy for something so obscure (unless it turns out not to be obscure). So I went with (3). Should we get a weaker typeclass lower in the lattice, we can replace this typeclass at no cost.

Reply:
Ok, then the current choice is very reasonable.